wled00/espnow_broadcast.cpp

```diff
@@ -0,0 +1,249 @@
+                        auto callback = _rxCallback;
+                        while( *callback ) {
+                            (*callback)(msg->mac, msg->data, msg->len);
+                            callback++;
+                        }
+                    }
+                    vRingbufferReturnItem(_networkRxMessages, (void *)msg);
+                } else {
+                    break;
+                }
+            }
+            break;
+        }
+        default:
+            break;
+    }
+}
+
+ESPNOWBroadcast::err_t ESPNOWBroadcast::send(const uint8_t* msg, size_t len) {
+    static const uint8_t broadcast[] = BROADCAST_ADDR_ARRAY_INITIALIZER;
+    return static_cast<ESPNOWBroadcast::err_t>(esp_now_send(broadcast, msg, len));
+}
+
+void ESPNOWBroadcast::start() {
+
+    Serial.println("starting ESPNow");
+
+    if ( WiFi.mode(WIFI_STA) ) {
+        if ( WiFi.status() == WL_DISCONNECTED ) {
+            if (esp_wifi_start() == ESP_OK) {
+                if (esp_now_init() == ESP_OK) {
+                    if (esp_now_register_recv_cb(onESPNowRxCallback) == ESP_OK) {
+                        static esp_now_peer_info_t peer = {
+                            BROADCAST_ADDR_ARRAY_INITIALIZER,
+                            {0},
+                            WLED_ESPNOW_WIFI_CHANNEL,
+                            WIFI_IF_STA,
+                            false,
+                            NULL
+                            };
+
+                        if (esp_now_add_peer(&peer) == ESP_OK) {
+                            ESPNOWBroadcast::STATE starting {ESPNOWBroadcast::STARTING};
+                            if (espnowBroadcast._state.compare_exchange_strong(starting, ESPNOWBroadcast::STARTED)) {
+                                return;
+                            } else {
+#ifdef NODE_DEBUGGING
+                                Serial.println("atomic state out of sync");
+#endif
+                            }
+                        } else {
+#ifdef NODE_DEBUGGING
+                            Serial.println("esp_now_add_peer failed");
+#endif
+                        }
+                    } else {
+#ifdef NODE_DEBUGGING
+                        Serial.println("esp_now_register_recv_cb failed");
+#endif
+                    }
+                } else {
+#ifdef NODE_DEBUGGING
+                    Serial.println("esp_now_init_init failed");
+#endif
+                }
+            } else {
+#ifdef NODE_DEBUGGING
+                Serial.println("esp_wifi_start failed");
+#endif
+            }
+        } else {
+#ifdef NODE_DEBUGGING
+            Serial.println("WiFi.status not disconnected");
+#endif
+        }
+    } else {
+#ifdef NODE_DEBUGGING
+        Serial.println("WiFi.mode failed");
+#endif
+    }
+    Serial.println("restarting ESPNow");
+    setupWiFi();
+}
+
+
+void ESPNOWBroadcast::onWiFiEvent(void* arg, esp_event_base_t event_base, int32_t event_id, void* event_data) {
+    if ( event_base == WIFI_EVENT ) {

```

Comment:
similar request for early exit on failure rather than nesting on success. But this style just personal preference

Reply:
I definitely prefer to not have a bunch of return paths.  makes multithreaded code more difficult to maintain

Reply:
I get it - fine to leave as is if you prefer it that way!

Just curious, why do you think it makes multi-threaded code more difficult to maintain?  the same issues can still happen where something was checked but changes values after the check
In my world of Javascript and Python I'm often not having to deal with multi-threaded the same way so would love to learn how this notation is better! for my own lessons :)

Reply:
When you start using critical sections and mutexes, you want to make sure the unwrap at the right times.

“using” in other languages help with this.   std::lock_guard() helps to, but it’s scope is less obvious

Reply:
If everything just does a simple return without a value or the same value, i can also see doing your pattern too. It just looked too messy with all the #ifdef’s